src/rules/typing/intent.rs

```diff
@@ -0,0 +1,153 @@
+use crate::ast::{child_with_name, to_text};
+use crate::settings::Settings;
+use crate::{ASTRule, Rule, Violation};
+use tree_sitter::Node;
+
+/// Rules for catching missing intent
+
+pub struct MissingIntent {}
+
+impl Rule for MissingIntent {
+    fn new(_settings: &Settings) -> Self {
+        MissingIntent {}
+    }
+
+    fn explain(&self) -> &'static str {
+        "
+        Procedure dummy arguments should have an explicit `intent`
+        attributes. This can help catch logic errors, potentially improve
+        performance, as well as serving as documentation for users of
+        the procedure.
+
+        Arguments with `intent(in)` are read-only input variables, and cannot be
+        modified by the routine.
+
+        Arguments with `intent(out)` are output variables, and their value on
+        entry into the routine can be safely ignored.
+
+        Finally, `intent(inout)` arguments can be both read and modified by the
+        routine. If an `intent` is not specified, it will default to
+        `intent(inout)`.
+        "
+    }
+}
+
+impl ASTRule for MissingIntent {
+    fn check<'a>(&self, node: &Node, src: &'a str) -> Option<Vec<Violation>> {
+        // Names of all the dummy arguments
+        let parameters: Vec<&str> = node
+            .child_by_field_name("parameters")?
+            .named_children(&mut node.walk())
+            .filter_map(|param| Some(to_text(&param, &src)?))
+            .collect();
+
+        let parent = node.parent()?;
+        let procedure_kind = parent.kind();
+
+        // Logic here is:
+        // 1. find variable declarations
+        // 2. filter to the declarations that don't have an `intent`
+        // 3. filter to the ones that contain any of the dummy arguments
+        // 4. collect into a vec of violations
+        //
+        // We filter by missing intent first, so we only have to
+        // filter by the dummy args once -- otherwise we either catch
+        // local var decls on the same line, or need to iterate over
+        // the decl names twice
+        let violations = parent
+            .named_children(&mut parent.walk())
+            .filter(|child| child.kind() == "variable_declaration")
+            .filter(|decl| {
+                !decl
+                    .children_by_field_name("attribute", &mut decl.walk())
+                    .any(|attr| {
+                        to_text(&attr, &src)
+                            .unwrap_or("")
+                            .to_lowercase()
+                            .starts_with("intent")
+                    })
+            })
+            .filter_map(|decl| {
```

Comment:
```suggestion
            .flat_map(|decl| {
```

Needs to be coupled with the next comment...